readthedocs/invitations/views.py

```diff
@@ -0,0 +1,129 @@
+"""Invitation views."""
+import structlog
+from django.contrib.auth.mixins import UserPassesTestMixin
+from django.http import Http404, HttpResponseBadRequest, HttpResponseRedirect
+from django.shortcuts import get_object_or_404
+from django.urls import reverse
+from django.views.generic import DeleteView, DetailView
+
+from readthedocs.core.mixins import PrivateViewMixin
+from readthedocs.core.permissions import AdminPermission
+from readthedocs.invitations.models import Invitation
+from readthedocs.organizations.models import Organization, Team, TeamInvite
+
+log = structlog.get_logger(__name__)
+
+
+class RevokeInvitation(PrivateViewMixin, UserPassesTestMixin, DeleteView):
+
+    model = Invitation
+    pk_url_kwarg = "invitation_pk"
+    http_method_names = ["post"]
+
+    def test_func(self):
+        invitation = self.get_object()
+        return invitation.can_revoke_invitation(self.request.user)
+
+    def get_success_url(self):
+        invitation = self.get_object()
+        return invitation.get_origin_url()
+
+
+class RedeemInvitation(DetailView):
+
+    """
+    Allow the user to accept or decline an invitation.
+
+    The invitation is deleted once it has been accepted or declined.
+    """
+
+    model = Invitation
```

Comment:
Perhaps add a `get_queryset` method to this view and make explicit what objects are eligible.

Reply:
All of them are valid, since we still show the page if the invitation is expired, but we do validate that expired invitations shouldn't be redeemed.

Reply:
Yes, I see now that this view perhaps will give more detailed outputs in the future and not just land people on a 404 page :+1: